torch/_library/custom_ops.py

```diff
@@ -0,0 +1,363 @@
+        >>> y = numpy_sin(x)
+        >>> assert torch.allclose(y, x.sin())
+        >>>
+        >>> # Example of a custom op that only works for one device type.
+        >>> @custom_op(mutated_args=(), types="cpu")
+        >>> def numpy_sin_cpu(x: Tensor) -> Tensor:
+        >>>     x_np = x.numpy()
+        >>>     y_np = np.sin(x_np)
+        >>>     return torch.from_numpy(y_np)
+        >>>
+        >>> x = torch.randn(3)
+        >>> y = numpy_sin_cpu(x)
+        >>> assert torch.allclose(y, x.sin())
+
+    """
+    assert len(mutated_args) == 0, "NYI"
+    assert qualname is None, "NYI"
+
+    def inner(fn):
+        return CustomOpDef._from_fn(fn, types, mutated_args)
+
+    return inner
+
+
+class CustomOpDef:
+    def __init__(self, namespace: str, name: str, schema: str, fn: Callable) -> None:
+        # Fields used to interface with the PyTorch dispatcher
+        self._namespace = namespace
+        self._name = name
+        self._qualname = f"{self._namespace}::{self._name}"
+        self._schema = schema
+
+        self._init_fn = fn
+
+        self._backend_fns: Dict[Union[str, None], Callable] = {}
+        self._abstract_fn: Optional[Callable] = None
+
+        self._lib = get_library_allowing_overwrite(self._namespace, self._name)
+        self._build()
+
+    @classmethod
+    def _from_fn(
+        cls,
+        fn: Callable,
+        types: device_types_t,
+        mutated_args: Sequence[str],
+    ) -> "CustomOpDef":
+        import torch
+
+        schema = torch._custom_op.impl.infer_schema(fn, mutated_args)
+        namespace = reserved_namespace()
+        name = utils.mangle(utils.unique_name(fn))
+        result = cls(namespace, name, schema, fn)
+        result.register_impl(types)(fn)
+        return result
+
+    def __repr__(self) -> str:
+        return f"<CustomOpDef({self._init_fn})>"
+
+    def register_impl(
+        self, types: device_types_t, fn: Optional[Callable] = None
+    ) -> Callable:
+        """Register an implementation for a device type for this operator.
+
+        Some valid types are: "cpu", "cuda", "xla", "mps", "ipu", "xpu".
+        This API may be used as a decorator.
+
+        Args:
+            types (str | Sequence[str]): The device types to register an impl to.
+
+        Examples::
+            >>> # xdoctest: +REQUIRES(env:TORCH_DOCTEST_CUDA)
+            >>> import torch
+            >>> from torch import Tensor
+            >>> from torch.library import custom_op
+            >>> import numpy as np
+            >>>
+            >>> # Example of split cpu and cuda definitions
+            >>> @custom_op(mutated_args=(), types="cpu")
+            >>> def numpy_sin(x: Tensor) -> Tensor:
+            >>>     x_np = x.numpy()
+            >>>     y_np = np.sin(x_np)
+            >>>     return torch.from_numpy(y_np)
+            >>>
+            >>> # Add implementations for the cuda device
+            >>> @numpy_sin.register_impl("cuda")
+            >>> def _(x):
```

Comment:
tbh I think this will lead to more questions than anything else.
I think we should use a name that makes sense here or numpy_sin like is done with python property.

Reply:
>  or numpy_sin like is done with python property.

What do you mean?

Reply:
The default example for property from python is to name all of these the same: https://docs.python.org/3/library/functions.html#property.deleter

Reply:
I need to go think about this a bit more

Reply:
We've got four options:
```py
@custom_op("mylib::sin", mutated_args=())
def sin(x):
    return x.clone()

# Option 1: recommend underscore name (The python singledispatch way)
# https://docs.python.org/3/library/functools.html#functools.singledispatch
@sin.register_fake
def _(x):
    return torch.empty_like(x)

# Option 2: recommend a named name
@sin.register_fake
def sin_fake(x):
    return torch.empty_like(x)

# Option 3: recommend the original name (The python property way)
@sin.register_fake
def sin(x):
    return torch.empty_like(x)

# Option 4: recommend a random name with convention
@sin.register_fake
def impl(x):
    return torch.empty_like(x)
```

I don't like Option 2 and Option 3 because those have bitten me in the past:
- Option 2 leads to unnecessarily verbose names and copy-paste confusion
- Option 3 means we can't call any of the impls directly.

Reply:
I don't really have a bone in this fight but Python singledispatch following the _ convention seems pretty compelling to me